File: backend/app/release_meta.py
import json
from pathlib import Path


RELEASE_META_FILENAME = ".release-meta.json"
PROJECT_ROOT = Path(__file__).resolve().parents[2]
REQUIRED_FIELDS = ("git_sha", "release_id", "built_at")
OPTIONAL_FIELDS = ("backend_slot", "n8n_compose_sha256")


def get_release_meta_path(project_root: Path | None = None) -> Path:
    return (project_root or PROJECT_ROOT) / RELEASE_META_FILENAME


def load_release_meta(path: Path | None = None) -> tuple[dict[str, str] | None, str | None]:
    release_meta_path = path or get_release_meta_path()

    if not release_meta_path.is_file():
        return None, "missing"

    try:
        payload = json.loads(release_meta_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None, "invalid_json"

    if not isinstance(payload, dict):
        return None, "invalid_shape"

    meta: dict[str, str] = {}
    for key in REQUIRED_FIELDS:
        value = payload.get(key)
        if not isinstance(value, str) or not value.strip():
            return None, "invalid_shape"
        meta[key] = value.strip()

    for key in OPTIONAL_FIELDS:
        value = payload.get(key)
        if isinstance(value, str) and value.strip():
            meta[key] = value.strip()

    return meta, None

File: frontend/lib/release-meta.ts
import { readFile } from "node:fs/promises";
import path from "node:path";


type ReleaseMetaPayload = {
  git_sha: string;
  release_id: string;
  built_at: string;
  backend_slot?: string;
  frontend_slot?: string;
};

export type RuntimeVersionPayload = {
  status: "ok" | "degraded";
  git_sha: string | null;
  release_id: string | null;
  built_at: string | null;
  frontend_slot: string | null;
  missing_release_meta: boolean;
  release_meta_error: string | null;
};


function getReleaseMetaPath() {
  return path.resolve(process.cwd(), "..", ".release-meta.json");
}


export async function readRuntimeVersion(): Promise<RuntimeVersionPayload> {
  const payload: RuntimeVersionPayload = {
    status: "ok",
    git_sha: null,
    release_id: null,
    built_at: null,
    frontend_slot: null,
    missing_release_meta: false,
    release_meta_error: null,
  };

  let parsed: unknown;
  try {
    parsed = JSON.parse(await readFile(getReleaseMetaPath(), "utf-8"));
  } catch (error) {
    payload.status = "degraded";
    payload.missing_release_meta = true;
    payload.release_meta_error = error instanceof SyntaxError ? "invalid_json" : "missing";
    return payload;
  }

  if (typeof parsed !== "object" || parsed === null) {
    payload.status = "degraded";
    payload.missing_release_meta = true;
    payload.release_meta_error = "invalid_shape";
    return payload;
  }

  const meta = parsed as Partial<ReleaseMetaPayload>;
  if (!meta.git_sha || !meta.release_id || !meta.built_at) {
    payload.status = "degraded";
    payload.missing_release_meta = true;
    payload.release_meta_error = "invalid_shape";
    return payload;
  }

  payload.git_sha = meta.git_sha;
  payload.release_id = meta.release_id;
  payload.built_at = meta.built_at;
  payload.frontend_slot = meta.frontend_slot ?? null;
  return payload;
}
